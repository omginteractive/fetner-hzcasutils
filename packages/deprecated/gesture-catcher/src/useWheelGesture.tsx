import {WheelGestureObservable} from '@hzdg/gesture-observable';
import useRefCallback from '@hzdg/use-ref-callback';
import {useObservableGestureEffect} from './utils';

export type WheelGestureObservableConfig = WheelGestureObservable.WheelGestureObservableConfig;
export type WheelGestureState = WheelGestureObservable.WheelGestureState;
export type WheelGestureEndState = WheelGestureObservable.WheelGestureEndState;
export type WheelGestureEventSourceState = WheelGestureObservable.WheelGestureEventSourceState;
export type WheelGestureConfig = Partial<
  WheelGestureObservable.WheelGestureObservableConfig
>;

/**
 * A callback for when wheel gesture state changes.
 * Receives either an ongoing gesture state, or the ending gesture state.
 */
export type WheelGestureChangeHandler = (
  state: WheelGestureState | WheelGestureEndState,
) => void;

export interface WheelGestureHandlers {
  /**
   * A callback for when a wheel gesture starts.
   * Receives the initializing gesture state.
   */
  onStart?: (state: WheelGestureState) => void;
  /**
   * A callback for when an ongoing wheel gesture updates.
   * Receives the latest gesture state.
   */
  onMove?: (state: WheelGestureState) => void;
  /**
   * A callback for when a wheel gesture has completed.
   * Receives the ending gesture state.
   */
  onEnd?: (state: WheelGestureEndState) => void;
}

export type WheelGestureHandler =
  | WheelGestureChangeHandler
  | WheelGestureHandlers;

/**
 * `useWheelGesture` is a React hook for components that want to handle
 * wheel gesture intent. It can be used with an existing ref or not.
 *
 * @see https://hz-core.netlify.app/use-wheel-gesture
 */
function useWheelGesture<T extends HTMLElement>(
  /**
   * An existing ref being passed to the DOM element on which to detect
   * wheel gestures. Useful for ref forwarding or sharing.
   */
  providedRef: React.RefObject<T>,
  /**
   * A function to handle wheel gesture updates,
   * or a configuration of handlers, like
   * `{onStart?, onMove?, onEnd?}`.
   */
  handler: WheelGestureHandler,
  /**
   * An object describing how to configure wheel gesture detection.
   */
  config?: WheelGestureConfig,
): void;
/**
 * `useWheelGesture` is a React hook for components that want to handle
 * wheel gesture intent. It can be used with an existing ref or not.
 *
 * @see https://hz-core.netlify.app/use-wheel-gesture
 */
function useWheelGesture<T extends HTMLElement>(
  /**
   * A function to handle wheel gesture updates,
   * or a configuration of handlers, like
   * `{onStart?, onMove?, onEnd?}`.
   */
  handler: WheelGestureHandler,
  /**
   * An object describing how to configure wheel gesture detection.
   */
  config?: WheelGestureConfig,
): (node: T | null) => void;
function useWheelGesture<T extends HTMLElement>(
  handlerOrProvidedRef: React.RefObject<T> | WheelGestureHandler,
  handlerOrConfig?: WheelGestureHandler | WheelGestureConfig,
  maybeConfig?: WheelGestureConfig,
): ((node: T | null) => void) | void {
  let handler: WheelGestureHandler;
  let config: WheelGestureConfig | undefined;
  let providedRef: React.RefObject<T> | null = null;

  if ('current' in handlerOrProvidedRef) {
    providedRef = handlerOrProvidedRef;
    handler = handlerOrConfig as WheelGestureHandler;
    config = maybeConfig;
  } else {
    handler = handlerOrProvidedRef as WheelGestureHandler;
    config = handlerOrConfig as WheelGestureConfig;
  }

  const [ref, setRef] = useRefCallback<T>(null);
  if (providedRef) setRef(providedRef.current);
  useObservableGestureEffect(WheelGestureObservable, ref, handler, config);
  if (!providedRef) return setRef;
}

export default useWheelGesture;
