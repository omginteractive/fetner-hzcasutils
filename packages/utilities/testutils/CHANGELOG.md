# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## 1.2.3 (2020-04-20)


### @hzdg/commitizen-config

* fix defaults when no config exists at all ([c31252e](https://github.com/hzdg/hz-core/commit/c31252e))


## 1.2.2 (2020-04-18)


* add netlify-preinstall script ([8b87c66](https://github.com/hzdg/hz-core/commit/8b87c66))


## [1.2.1](https://github.com/hzdg/hz-core/compare/testutils@1.2.0...testutils@1.2.1) (2020-04-15)


* namespace under @hzdg ([b1b92f3](https://github.com/hzdg/hz-core/commit/b1b92f3))


# [1.2.0](https://github.com/hzdg/hz-core/compare/testutils@1.1.0...testutils@1.2.0) (2020-02-24)


### testutils

* add KeyboardSequence modifier methods ([0b7b50e](https://github.com/hzdg/hz-core/commit/0b7b50e))
* fix EventSequence run queue continuation ([c217066](https://github.com/hzdg/hz-core/commit/c217066))
* add return types ([f487679](https://github.com/hzdg/hz-core/commit/f487679))


# [1.1.0](https://github.com/hzdg/hz-core/compare/testutils@1.0.1...testutils@1.1.0) (2020-02-10)


### testutils

* add repeat() to {Mouse,Touch,Wheel}Sequence ([b0ebe9a](https://github.com/hzdg/hz-core/commit/b0ebe9a))
* fix getComputedValue overload types ([bfc6031](https://github.com/hzdg/hz-core/commit/bfc6031))


## [1.0.1](https://github.com/hzdg/hz-core/compare/testutils@1.0.0...testutils@1.0.1) (2019-11-08)


### testutils

* move MockMutationObserver to testutils ([a87950f](https://github.com/hzdg/hz-core/commit/a87950f))
* fix type errors ([0743e7a](https://github.com/hzdg/hz-core/commit/0743e7a))


## [1.0.0](https://github.com/hzdg/hz-core/compare/testutils@0.0.2...testutils@1.0.0) (2019-05-20)


### testutils

* fix EventSequence API and types ([2317f2f](https://github.com/hzdg/hz-core/commit/2317f2f))


## 0.0.2 (2019-05-03)


### testutils

* move testutils to their own package ([22ba155](https://github.com/hzdg/hz-core/commit/22ba155))
