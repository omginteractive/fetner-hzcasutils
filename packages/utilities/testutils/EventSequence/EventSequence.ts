/* eslint-env jest */

interface Extensions {
  [key: string]: Function | false | null;
}

/**
 * Pick only the members of base type `B`
 * that extend the predicate type `T`.
 */
type PickType<B, T> = Pick<
  B,
  {[K in keyof B]: B[K] extends T ? K : never}[keyof B]
>;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type Newable = new (...args: any[]) => any;

export default class EventSequence<
  EventType extends Event = Event,
  EventInitType extends EventInit = EventInit
> implements PromiseLike<EventType[]> {
  constructor(public node: HTMLElement | Document) {}

  static create<T extends Newable>(
    this: T,
    ...args: ConstructorParameters<T>
  ): InstanceType<T> {
    return new this(...args);
  }

  eventQueue: [string, EventInitType?][] = [];
  dispatched: EventType[] = [];

  createNextEvent(
    type: string,
    init?: EventInitType,
    lastEvent?: EventType, // eslint-disable-line @typescript-eslint/no-unused-vars
  ): EventType {
    return new Event(type, init) as EventType;
  }

  normalizeEventInit(
    type: string,
    init?: EventInitType,
  ): EventInitType | undefined {
    return init;
  }

  dispatch<T extends EventSequence<EventType, EventInitType>>(
    this: T,
    type: string,
    init?: EventInitType,
  ): T {
    this.eventQueue.push([type, this.normalizeEventInit(type, init)]);
    return this;
  }

  expose<
    T extends EventSequence<EventType, EventInitType>,
    E extends Extensions
  >(
    this: T,
    extensions: E,
  ): Pick<T, Exclude<keyof T, keyof E>> & PickType<E, Function> {
    const extended = Object.assign(Object.create(this), extensions);
    for (const key in extensions) {
      if (key === 'dispatch' || key === 'expose' || key === 'run') continue;
      if (!Object.prototype.hasOwnProperty.call(extensions, key)) continue;
      if (!extensions[key]) extended[key] = undefined;
    }
    return extended;
  }

  run(): Promise<EventType[]> {
    return this.eventQueue.splice(0, this.eventQueue.length).reduce(
      (lastEvent, [eventType, eventInit]) =>
        lastEvent.then(async dispatched => {
          const nextEvent = this.createNextEvent(
            eventType,
            eventInit,
            dispatched[dispatched.length - 1],
          ) as EventType;
          this.node.dispatchEvent(nextEvent);
          dispatched.push(nextEvent);
          return dispatched;
        }),
      Promise.resolve(this.dispatched),
    );
  }

  async then<Result = EventType[], Reason = never>(
    resolve: (value: EventType[]) => Result | PromiseLike<Result>,
    reject?: ((reason: unknown) => Reason | PromiseLike<Reason>) | null,
  ): Promise<Result | Reason> {
    try {
      const result = await this.run();
      return resolve(result);
    } catch (error) {
      if (reject) return reject(error);
      throw error;
    }
  }
}
