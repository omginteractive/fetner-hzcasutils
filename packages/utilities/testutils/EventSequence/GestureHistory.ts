export default class GestureHistory<T>
  implements ZenObservable.Observer<T>, Iterable<T> {
  history: T[] = [];
  _error: Error | null = null;
  _complete = false;
  next = (value: T): void => {
    this.history.push(value);
  };
  error = (error: Error): void => {
    this._error = error;
  };
  complete = (): void => {
    this._complete = true;
  };
  nth(i: number): T | null {
    return this.history[i - 1];
  }
  get size(): number {
    return this.history.length;
  }
  get first(): T | null {
    return this.history[0];
  }
  get last(): T | null {
    return this.history[this.history.length - 1];
  }
  *[Symbol.iterator](): Iterator<T> {
    let i = 0;
    while (i < this.history.length) {
      yield this.history[i++];
    }
  }
}
