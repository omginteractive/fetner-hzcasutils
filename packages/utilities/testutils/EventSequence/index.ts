/* eslint-env jest */
export {default, default as EventSequence} from './EventSequence';
export {default as MouseSequence} from './MouseSequence';
export {default as TouchSequence} from './TouchSequence';
export {default as KeyboardSequence} from './KeyboardSequence';
export {default as WheelSequence} from './WheelSequence';
export {default as GestureHistory} from './GestureHistory';
