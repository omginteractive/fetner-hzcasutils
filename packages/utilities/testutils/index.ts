export {default as sleep} from './sleep';
export {default as hasProperty} from './hasProperty';
export {default as getFlag} from './getFlag';
export {default as getValue} from './getValue';
export {default as getTouchAt} from './getTouchAt';
export {default as getComputedValue} from './getComputedValue';
