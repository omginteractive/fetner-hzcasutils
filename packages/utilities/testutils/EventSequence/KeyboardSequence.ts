/* eslint-env jest */
import {
  KEY_DOWN,
  KEY_UP,
  ENTER,
  TAB,
  SPACE,
  PAGE_UP,
  PAGE_DOWN,
  END,
  HOME,
  ARROW_LEFT,
  ARROW_UP,
  ARROW_RIGHT,
  ARROW_DOWN,
  BACKSPACE,
  ESCAPE,
  CODES,
  CODES_2_KEY_CODES,
} from './types';
import EventSequence from './EventSequence';
import {getFlag, getValue} from '..';

type KeyboardEventType = typeof KEY_DOWN | typeof KEY_UP;

type UnnormalizedKeyboardEventInit = KeyboardEventInit & {
  keyCode?: number;
  which?: number;
  charCode?: number;
};

type KeyboardDownEventInit = UnnormalizedKeyboardEventInit & {
  key: string;
};

type Omit<T, K> = Pick<T, Exclude<keyof T, K>>;

interface Modifiers {
  ctrlKey: boolean;
  altKey: boolean;
  metaKey: boolean;
  shiftKey: boolean;
}

interface KeyboardDownSequence
  extends Omit<
    KeyboardSequence,
    'down' | 'modify' | 'ctrl' | 'alt' | 'meta' | 'shift'
  > {
  repeat(count?: number): KeyboardDownSequence;
  modify(modifiers: Partial<Modifiers>): KeyboardDownSequence;
  ctrl(active?: boolean): KeyboardDownSequence;
  alt(active?: boolean): KeyboardDownSequence;
  meta(active?: boolean): KeyboardDownSequence;
  shift(active?: boolean): KeyboardDownSequence;
  up(): KeyboardSequence;
}

function getCode(key: string): string {
  if (key === ' ') return SPACE;
  if (CODES.includes(key)) return key;
  return `key${key.toUpperCase()}`;
}

function getKeyCode(key: string): number {
  const code = getCode(key);
  if (CODES.includes(code)) {
    return CODES_2_KEY_CODES[code];
  }
  return key.charCodeAt(0);
}

function normalizeKeyboardEventInit(
  init: UnnormalizedKeyboardEventInit,
  from: KeyboardEventInit,
): EventInit & UnnormalizedKeyboardEventInit {
  const key = getValue(init, 'key', from.key);
  return {
    key,
    code: key ? getCode(key) : undefined,
    keyCode: key ? getKeyCode(key) : undefined,
    which: key ? getKeyCode(key) : undefined,
    charCode: 0,
    ctrlKey: getFlag(init, 'ctrlKey', from.ctrlKey),
    shiftKey: getFlag(init, 'shiftKey', from.shiftKey),
    altKey: getFlag(init, 'altKey', from.altKey),
    metaKey: getFlag(init, 'metaKey', from.metaKey),
    location: getValue(init, 'location', from.location),
    repeat: getFlag(init, 'repeat', from.repeat),
    bubbles: true,
    cancelable: true,
    view: window,
  };
}
export default class KeyboardSequence extends EventSequence<
  KeyboardEvent,
  UnnormalizedKeyboardEventInit
> {
  modifiers: Modifiers = {
    ctrlKey: false,
    altKey: false,
    metaKey: false,
    shiftKey: false,
  };

  createNextEvent(
    type: KeyboardEventType,
    init: UnnormalizedKeyboardEventInit = {},
    lastEvent?: KeyboardEvent | undefined,
  ): KeyboardEvent {
    return new KeyboardEvent(
      type,
      normalizeKeyboardEventInit(init, lastEvent || new KeyboardEvent('init')),
    );
  }
  normalizeEventInit(
    type: KeyboardEventType,
    init: UnnormalizedKeyboardEventInit = {},
  ): UnnormalizedKeyboardEventInit {
    return {...this.modifiers, ...init};
  }
  down(downOpts: KeyboardDownEventInit): KeyboardDownSequence {
    const downSequence: KeyboardDownSequence = this.dispatch(
      KEY_DOWN,
      downOpts,
    ).expose({
      down: false,
      ctrl(active = true): KeyboardDownSequence {
        return downSequence.modify({ctrlKey: active});
      },
      alt(active = true): KeyboardDownSequence {
        return downSequence.modify({altKey: active});
      },
      meta(active = true): KeyboardDownSequence {
        return downSequence.modify({metaKey: active});
      },
      shift(active = true): KeyboardDownSequence {
        return downSequence.modify({shiftKey: active});
      },
      modify: (modifiers: Partial<Modifiers>): KeyboardDownSequence => {
        const lastEvent = this.eventQueue[this.eventQueue.length - 1];
        if (lastEvent && lastEvent[0] !== KEY_UP) {
          this.modify(modifiers);
        } else {
          this.modifiers = {...this.modifiers, ...modifiers};
          const lastDispatched = this.dispatched[this.dispatched.length - 1];
          if (lastDispatched && lastDispatched.type !== KEY_UP) {
            this._dispatchModificationIfNecessary(
              lastDispatched.type,
              lastDispatched,
            );
          }
        }
        return downSequence;
      },
      repeat(count = 1): KeyboardDownSequence {
        if (count <= 0) throw new Error('count must be a positive integer!');
        let d = downSequence;
        for (let i = 0; i < count; i++) {
          d = downSequence.dispatch(KEY_DOWN, {repeat: true});
        }
        return d;
      },
      up: (): KeyboardSequence => this.dispatch(KEY_UP, {repeat: false}),
    });
    return downSequence;
  }
  space(): KeyboardDownSequence {
    return this.down({key: ' '});
  }
  pageUp(): KeyboardDownSequence {
    return this.down({key: PAGE_UP});
  }
  pageDown(): KeyboardDownSequence {
    return this.down({key: PAGE_DOWN});
  }
  end(): KeyboardDownSequence {
    return this.down({key: END});
  }
  home(): KeyboardDownSequence {
    return this.down({key: HOME});
  }
  arrowLeft(): KeyboardDownSequence {
    return this.down({key: ARROW_LEFT});
  }
  arrowUp(): KeyboardDownSequence {
    return this.down({key: ARROW_UP});
  }
  arrowRight(): KeyboardDownSequence {
    return this.down({key: ARROW_RIGHT});
  }
  arrowDown(): KeyboardDownSequence {
    return this.down({key: ARROW_DOWN});
  }
  enter(): KeyboardDownSequence {
    return this.down({key: ENTER});
  }
  tab(): KeyboardDownSequence {
    return this.down({key: TAB});
  }
  backspace(): KeyboardDownSequence {
    return this.down({key: BACKSPACE});
  }
  escape(): KeyboardDownSequence {
    return this.down({key: ESCAPE});
  }
  private _dispatchModificationIfNecessary(
    type: string,
    init: KeyboardEventInit,
  ): KeyboardSequence {
    if (
      Boolean(init.ctrlKey) !== this.modifiers.ctrlKey ||
      Boolean(init.altKey) !== this.modifiers.altKey ||
      Boolean(init.metaKey) !== this.modifiers.metaKey ||
      Boolean(init.shiftKey) !== this.modifiers.shiftKey
    ) {
      return this.dispatch(
        type,
        normalizeKeyboardEventInit(this.modifiers, init),
      );
    }
    return this;
  }
  modify(modifiers: Partial<Modifiers>): KeyboardSequence {
    this.modifiers = {...this.modifiers, ...modifiers};
    const lastEvent = this.eventQueue[this.eventQueue.length - 1];
    if (lastEvent && lastEvent[0] !== KEY_UP && lastEvent[1]) {
      return this._dispatchModificationIfNecessary(lastEvent[0], lastEvent[1]);
    }
    return this;
  }
  ctrl(active = true): KeyboardSequence {
    return this.modify({ctrlKey: active});
  }
  alt(active = true): KeyboardSequence {
    return this.modify({altKey: active});
  }
  meta(active = true): KeyboardSequence {
    return this.modify({metaKey: active});
  }
  shift(active = true): KeyboardSequence {
    return this.modify({shiftKey: active});
  }
}
