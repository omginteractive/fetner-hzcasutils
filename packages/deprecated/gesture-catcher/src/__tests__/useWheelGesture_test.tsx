/** @jsdom-global Object */
/* eslint-env jest, browser */
import React from 'react';
import {render, act} from '@testing-library/react';
import {WheelSequence} from 'testutils/EventSequence';
import useWheelGesture, {WheelGestureConfig} from '../useWheelGesture';

const GestureStateMatcher = {
  duration: expect.any(Number),
  elapsed: expect.any(Number),
  time: expect.any(Number),
  timeInitial: expect.any(Number),
};

// Threshold required to trigger a wheel gesture.
const DEFAULT_THRESHOLD = 40;

beforeEach(() => {
  jest.useFakeTimers();
});

function runAllTimers(): void {
  act(() => {
    jest.runAllTimers();
  });
}

test('useWheelGesture passes gesture states to a handler', async () => {
  const handler = jest.fn();
  const WheelGestureUser = (): JSX.Element => {
    const ref = useWheelGesture<HTMLDivElement>(handler);
    return <div ref={ref} data-testid="gesture-target" />;
  };
  const {getByTestId} = render(<WheelGestureUser />);
  const target = getByTestId('gesture-target');
  await new WheelSequence(target)
    .wheel({deltaX: DEFAULT_THRESHOLD + 1})
    .wheel({deltaX: 5})
    .wheel({deltaX: 3, deltaY: 5});
  runAllTimers();
  expect(handler.mock.calls).toHaveLength(4);
  handler.mock.calls.forEach(([state]) => {
    expect(state).toMatchSnapshot(GestureStateMatcher);
  });
});

test('useWheelGesture uses an existing ref', async () => {
  const handler = jest.fn();
  const WheelGestureUser = (): JSX.Element => {
    const [node, setNode] = React.useState<HTMLDivElement | null>(null);
    const ref = React.useRef(node);
    ref.current = node;
    useWheelGesture<HTMLDivElement>(ref, handler);
    return <div ref={node => setNode(node)} data-testid="gesture-target" />;
  };
  const {getByTestId} = render(<WheelGestureUser />);
  const target = getByTestId('gesture-target');
  await new WheelSequence(target).wheel({deltaY: DEFAULT_THRESHOLD + 5});
  runAllTimers();
  expect(handler.mock.calls).toHaveLength(2);
  handler.mock.calls.forEach(([state]) => {
    expect(state).toMatchSnapshot(GestureStateMatcher);
  });
});

test('useWheelGesture handles changes to handler and ref', async () => {
  // Create two handlers.
  const handler1 = jest.fn();
  const handler2 = jest.fn();

  // Create two refs.
  const ref1 = React.createRef();
  const ref2 = React.createRef();

  const WheelGestureUser = React.forwardRef(
    ({handler}: {handler: jest.Mock}, forwardedRef): JSX.Element => {
      // Use a different ref based on the index prop.
      const [node, setNode] = React.useState<HTMLDivElement | null>(null);
      const ref = forwardedRef as React.MutableRefObject<HTMLDivElement | null>;
      ref.current = node;
      useWheelGesture<HTMLDivElement>(ref, handler);
      return (
        <div>
          <div
            ref={ref === ref1 ? node => setNode(node) : null}
            data-testid="gesture-target1"
          />
          <div
            ref={ref === ref2 ? node => setNode(node) : null}
            data-testid="gesture-target2"
          />
        </div>
      );
    },
  );
  WheelGestureUser.displayName = 'WheelGestureUser';

  const {getByTestId, rerender} = render(
    <WheelGestureUser handler={handler1} ref={ref1} />,
  );

  const target1 = getByTestId('gesture-target1');
  const target2 = getByTestId('gesture-target2');

  // Test that gesturing on the first target updates the current handler.
  await new WheelSequence(target1).wheel({deltaY: DEFAULT_THRESHOLD + 3});
  runAllTimers();
  expect(handler1.mock.calls).toHaveLength(2);
  expect(handler2.mock.calls).toHaveLength(0);

  handler1.mockReset();
  handler2.mockReset();

  // Test that gesturing on the second target
  // does not update the current handler.
  await new WheelSequence(target2).wheel({deltaY: DEFAULT_THRESHOLD + 3});
  runAllTimers();
  expect(handler1.mock.calls).toHaveLength(0);
  expect(handler2.mock.calls).toHaveLength(0);

  // Switch the ref to the other gesture target.
  rerender(<WheelGestureUser handler={handler1} ref={ref2} />);

  // Test that gesturing on the first target
  // does not update the current handler.
  await new WheelSequence(target1).wheel({deltaY: DEFAULT_THRESHOLD + 3});
  runAllTimers();
  expect(handler1.mock.calls).toHaveLength(0);
  expect(handler2.mock.calls).toHaveLength(0);

  handler1.mockReset();
  handler2.mockReset();

  // Test that gesturing on the second target updates the current handler.
  await new WheelSequence(target2).wheel({deltaY: DEFAULT_THRESHOLD + 3});
  runAllTimers();
  expect(handler1.mock.calls).toHaveLength(2);
  expect(handler2.mock.calls).toHaveLength(0);

  handler1.mockReset();
  handler2.mockReset();

  // Switch the handler.
  rerender(<WheelGestureUser handler={handler2} ref={ref2} />);

  // Test that gesturing on the first target
  // does not update the current handler.
  await new WheelSequence(target1).wheel({deltaY: DEFAULT_THRESHOLD + 3});
  runAllTimers();
  expect(handler1.mock.calls).toHaveLength(0);
  expect(handler2.mock.calls).toHaveLength(0);

  handler1.mockReset();
  handler2.mockReset();

  // Test that gesturing on the second target updates the current handler.
  await new WheelSequence(target2).wheel({deltaY: DEFAULT_THRESHOLD + 3});
  runAllTimers();
  expect(handler1.mock.calls).toHaveLength(0);
  expect(handler2.mock.calls).toHaveLength(2);

  handler1.mockReset();
  handler2.mockReset();

  // Switch the ref back to the first target.
  rerender(<WheelGestureUser handler={handler2} ref={ref1} />);

  // Test that gesturing on the first target updates the current handler.
  await new WheelSequence(target1).wheel({deltaY: DEFAULT_THRESHOLD + 3});
  runAllTimers();
  expect(handler1.mock.calls).toHaveLength(0);
  expect(handler2.mock.calls).toHaveLength(2);

  handler1.mockReset();
  handler2.mockReset();

  // Test that gesturing on the second target
  // does not update the current handler.
  await new WheelSequence(target2).wheel({deltaY: DEFAULT_THRESHOLD + 3});
  runAllTimers();
  expect(handler1.mock.calls).toHaveLength(0);
  expect(handler2.mock.calls).toHaveLength(0);
});

test('useWheelGesture resubscribes on config changes', async () => {
  const handler = jest.fn();
  const WheelGestureUser = ({
    config,
  }: {
    config?: WheelGestureConfig;
  }): JSX.Element => {
    const ref = useWheelGesture<HTMLDivElement>(handler, config);
    return <div ref={ref} data-testid="gesture-target" />;
  };
  const {getByTestId, rerender} = render(
    <WheelGestureUser config={{threshold: 9}} />,
  );
  const target = getByTestId('gesture-target');
  await new WheelSequence(target).wheel({deltaX: 1}).repeat(8);
  runAllTimers();
  expect(handler.mock.calls).toHaveLength(0);

  handler.mockReset();

  rerender(<WheelGestureUser config={{threshold: 10}} />);
  await new WheelSequence(target).wheel({deltaX: 1}).repeat(10);
  runAllTimers();
  expect(handler.mock.calls).toHaveLength(2);
  handler.mock.calls.forEach(([state]) => {
    expect(state).toMatchSnapshot(GestureStateMatcher, 'threshold 10');
  });

  handler.mockReset();

  rerender(<WheelGestureUser />);
  await new WheelSequence(target)
    .wheel({deltaX: DEFAULT_THRESHOLD + 1})
    .wheel({deltaX: 1});
  runAllTimers();
  expect(handler.mock.calls).toHaveLength(3);
  handler.mock.calls.forEach(([state]) => {
    expect(state).toMatchSnapshot(GestureStateMatcher, 'no config');
  });
});
