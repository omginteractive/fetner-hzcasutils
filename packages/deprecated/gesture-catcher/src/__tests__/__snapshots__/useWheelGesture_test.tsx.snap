// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useWheelGesture passes gesture states to a handler 1`] = `
Object {
  "duration": Any<Number>,
  "elapsed": Any<Number>,
  "gesturing": true,
  "time": Any<Number>,
  "timeInitial": Any<Number>,
  "type": "wheel",
  "x": 0,
  "xDelta": -41,
  "xInitial": 0,
  "xPrev": 0,
  "xSpin": 1,
  "xVelocity": -41,
  "y": 0,
  "yDelta": 0,
  "yInitial": 0,
  "yPrev": 0,
  "ySpin": 0,
  "yVelocity": 0,
}
`;

exports[`useWheelGesture passes gesture states to a handler 2`] = `
Object {
  "duration": Any<Number>,
  "elapsed": Any<Number>,
  "gesturing": true,
  "time": Any<Number>,
  "timeInitial": Any<Number>,
  "type": "wheel",
  "x": 0,
  "xDelta": -46,
  "xInitial": 0,
  "xPrev": 0,
  "xSpin": 2,
  "xVelocity": -5,
  "y": 0,
  "yDelta": 0,
  "yInitial": 0,
  "yPrev": 0,
  "ySpin": 0,
  "yVelocity": 0,
}
`;

exports[`useWheelGesture passes gesture states to a handler 3`] = `
Object {
  "duration": Any<Number>,
  "elapsed": Any<Number>,
  "gesturing": true,
  "time": Any<Number>,
  "timeInitial": Any<Number>,
  "type": "wheel",
  "x": 0,
  "xDelta": -49,
  "xInitial": 0,
  "xPrev": 0,
  "xSpin": 3,
  "xVelocity": -3,
  "y": 0,
  "yDelta": -5,
  "yInitial": 0,
  "yPrev": 0,
  "ySpin": 1,
  "yVelocity": -5,
}
`;

exports[`useWheelGesture passes gesture states to a handler 4`] = `
Object {
  "duration": Any<Number>,
  "elapsed": Any<Number>,
  "gesturing": false,
  "time": Any<Number>,
  "timeInitial": Any<Number>,
  "type": "gestureend",
  "x": 0,
  "xDelta": -49,
  "xInitial": 0,
  "xPrev": 0,
  "xSpin": 3,
  "xVelocity": -3,
  "y": 0,
  "yDelta": -5,
  "yInitial": 0,
  "yPrev": 0,
  "ySpin": 1,
  "yVelocity": -5,
}
`;

exports[`useWheelGesture resubscribes on config changes: no config 1`] = `
Object {
  "duration": Any<Number>,
  "elapsed": Any<Number>,
  "gesturing": true,
  "time": Any<Number>,
  "timeInitial": Any<Number>,
  "type": "wheel",
  "x": 0,
  "xDelta": -41,
  "xInitial": 0,
  "xPrev": 0,
  "xSpin": 1,
  "xVelocity": -41,
  "y": 0,
  "yDelta": 0,
  "yInitial": 0,
  "yPrev": 0,
  "ySpin": 0,
  "yVelocity": 0,
}
`;

exports[`useWheelGesture resubscribes on config changes: no config 2`] = `
Object {
  "duration": Any<Number>,
  "elapsed": Any<Number>,
  "gesturing": true,
  "time": Any<Number>,
  "timeInitial": Any<Number>,
  "type": "wheel",
  "x": 0,
  "xDelta": -42,
  "xInitial": 0,
  "xPrev": 0,
  "xSpin": 2,
  "xVelocity": -1,
  "y": 0,
  "yDelta": 0,
  "yInitial": 0,
  "yPrev": 0,
  "ySpin": 0,
  "yVelocity": 0,
}
`;

exports[`useWheelGesture resubscribes on config changes: no config 3`] = `
Object {
  "duration": Any<Number>,
  "elapsed": Any<Number>,
  "gesturing": false,
  "time": Any<Number>,
  "timeInitial": Any<Number>,
  "type": "gestureend",
  "x": 0,
  "xDelta": -42,
  "xInitial": 0,
  "xPrev": 0,
  "xSpin": 2,
  "xVelocity": -1,
  "y": 0,
  "yDelta": 0,
  "yInitial": 0,
  "yPrev": 0,
  "ySpin": 0,
  "yVelocity": 0,
}
`;

exports[`useWheelGesture resubscribes on config changes: threshold 10 1`] = `
Object {
  "duration": Any<Number>,
  "elapsed": Any<Number>,
  "gesturing": true,
  "time": Any<Number>,
  "timeInitial": Any<Number>,
  "type": "wheel",
  "x": 0,
  "xDelta": -1,
  "xInitial": 0,
  "xPrev": 0,
  "xSpin": 1,
  "xVelocity": -1,
  "y": 0,
  "yDelta": 0,
  "yInitial": 0,
  "yPrev": 0,
  "ySpin": 0,
  "yVelocity": 0,
}
`;

exports[`useWheelGesture resubscribes on config changes: threshold 10 2`] = `
Object {
  "duration": Any<Number>,
  "elapsed": Any<Number>,
  "gesturing": false,
  "time": Any<Number>,
  "timeInitial": Any<Number>,
  "type": "gestureend",
  "x": 0,
  "xDelta": -1,
  "xInitial": 0,
  "xPrev": 0,
  "xSpin": 1,
  "xVelocity": -1,
  "y": 0,
  "yDelta": 0,
  "yInitial": 0,
  "yPrev": 0,
  "ySpin": 0,
  "yVelocity": 0,
}
`;

exports[`useWheelGesture uses an existing ref 1`] = `
Object {
  "duration": Any<Number>,
  "elapsed": Any<Number>,
  "gesturing": true,
  "time": Any<Number>,
  "timeInitial": Any<Number>,
  "type": "wheel",
  "x": 0,
  "xDelta": 0,
  "xInitial": 0,
  "xPrev": 0,
  "xSpin": 0,
  "xVelocity": 0,
  "y": 0,
  "yDelta": -45,
  "yInitial": 0,
  "yPrev": 0,
  "ySpin": 1,
  "yVelocity": -45,
}
`;

exports[`useWheelGesture uses an existing ref 2`] = `
Object {
  "duration": Any<Number>,
  "elapsed": Any<Number>,
  "gesturing": false,
  "time": Any<Number>,
  "timeInitial": Any<Number>,
  "type": "gestureend",
  "x": 0,
  "xDelta": 0,
  "xInitial": 0,
  "xPrev": 0,
  "xSpin": 0,
  "xVelocity": 0,
  "y": 0,
  "yDelta": -45,
  "yInitial": 0,
  "yPrev": 0,
  "ySpin": 1,
  "yVelocity": -45,
}
`;
